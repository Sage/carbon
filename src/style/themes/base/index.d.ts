import { Palette } from "../../palette";

interface ColorsWithHex {
  [key: string]: string;
}

type BasePalette = Palette & ColorsWithHex & {
  blackOpacity: (opacity: number) => string;
  whiteOpacity: (opacity: number) => string;
};

interface Colors {
  base: string;
  primary: string;
  secondary: string;
  tertiary: string;
  brand: string;
  disabled: string;
  whiteMix: string;
  withOpacity: string;

  // generic
  black: "#000000";
  slate: "#003349";
  white: "#FFFFFF";

  // element
  border: string;
  dashedBorder: string;
  dashedButtonText: string;
  dashedHoverBackground: string;
  focusedIcon: string;
  focusedLinkBackground: string;
  previewBackground: string;
  hoveredTabKeyline: string;

  // status
  error: string;
  focus: string;
  info: string;
  success: string;
  warning: string;
  destructive: {
    hover: string;
  };
  asterisk: string;
}

export interface ThemeObject {
  name: string;
  palette: BasePalette;
  spacing: number;
  space: number[];

  colors: Colors;

  anchorNavigation: {
    divider: string;
    navItemHoverBackground: string;
  };

  accordion: {
    border: string;
    background: string;
  };

  tileSelect: {
    border: string;
    disabledBackground: string;
    hoverBackground: string;
    descriptionColor: string;
    disabledText: string;
  };

  batchSelection: {
    lightTheme: string;
  };

  hr: {
    background: string;
  };

  editor: {
    border: string;
    counter: string;
    placeholder: string;
    button: {
      hover: string;
    };
    toolbar: {
      background: string;
    };
  };

  menu: {
    focus: string;
    divider: string;
    itemColor: string;
    itemColorDisabled: string;

    light: {
      background: string;
      selected: string;
      divider: string;
      title: string;
    };

    dark: {
      divider: string;
      submenuBackground: string;
      selected: string;
      title: string;
      searchIcon: string;
      searchIconHover: string;
    };
  };

  form: {
    invalid: string;
  };

  card: {
    footerBackground: string;
    footerBorder: string;
  };

  carousel: {
    activeSelectorBackground: string;
    inactiveSelectorBackground: string;
  };

  flatTable: {
    light: {
      headerBackground: string;
      border: string;
    };

    dark: {
      headerBackground: string;
      border: string;
    };

    subRow: {
      background: string;
      shadow: string;
    };

    transparentWhite: {
      headerBackground: string;
      border: string;
    };

    transparentBase: {
      headerBackground: string;
      border: string;
    };

    drawerSidebar: {
      headerBackground: string;
      hover: string;
      highlighted: string;
      selected: string;
    };

    hover: string;
    headerIconColor: string;
    selected: string;
    highlighted: string;
  };

  help: {
    color: string;
    hover: string;
  };

  pod: {
    border: string;
    secondaryBackground: string;
    tertiaryBackground: string;
    tileBackground: string;
    footerBackground: string;
    hoverBackground: string;
  };

  text: {
    color: string;
    placeholder: string;
    size: string;
  };

  readOnly: {
    textboxBackground: string;
    textboxBorder: string;
    textboxText: string;
  };

  content: {
    secondaryColor: string;
  };

  definitionList: {
    dtTextDark: string;
    dtTextLight: string;
    ddText: string;
  };

  disabled: {
    border: string;
    button: string;
    disabled: string;
    input: string;
    text: string;
    buttonText: string;
    background: string;
    switch: string;
  };

  draggableItem: {
    border: string;
  };

  checkable: {
    checked: string;
  };

  table: {
    primary: string;
    secondary: string;
    tertiary: string;
    header: string;
    hover: string;
    selected: string;
    zebra: string;
    dragging: string;
  };

  drawer: {
    background: string;
    divider: string;
  };

  pager: {
    active: string;
    disabled: string;
    alternate: string;
  };

  icon: {
    default: string;
    defaultHover: string;
    onLightBackground: string;
    onLightBackgroundHover: string;
    disabled: string;
  };

  popoverContainer: {
    iconColor: string;
  };

  navigationBar: {
    light: {
      background: string;
      borderBottom: string;
    };

    dark: {
      background: string;
      borderBottom: string;
    };
  };

  numeralDate: {
    passive: string;
    error: string;
  };

  portrait: {
    border: string;
    background: string;
    initials: string;
  };

  picklist: {
    locked: string;
    lockedContent: string;
    lockedText: string;
  };

  pill: {
    neutral: string;
    warning: string;
    neutralBackgroundFocus: string;
    warningButtonFocus: string;
    errorButtonFocus: string;
  };

  scrollbar: {
    light: {
      thumb: string;
      track: string;
    };

    dark: {
      thumb: string;
      track: string;
    };
  };

  search: {
    active: string;
    button: string;
    passive: string;
    icon: string;
    iconHover: string;
    searchActive: string;
    darkVariantPlaceholder: string;
    darkVariantBorder: string;
    darkVariantText: string;
    iconDarkVariant: string;
    iconDarkVariantHover: string;
  };

  select: {
    border: string;
    selected: string;
    optionHeader: string;
    tableHeaderBorder: string;
  };

  shadows: {
    depth1: string;
    depth2: string;
    depth3: string;
    depth4: string;
    cards: string;
    cardsIE: string;
  };

  switch: {
    off: string;
  };

  tile: {
    border: string;
    footerBackground: string;
    separator: string;
  };

  tab: {
    background: string;
    altHover: string;
  };

  note: {
    timeStamp: string;
  };

  zIndex: {
    smallOverlay: number;
    overlay: number;
    popover: number;
    modal: number;
    header: number;
    fullScreenModal: number;
    notification: number;
    aboveAll: number;
  };
}

declare const baseTheme: ThemeObject;

declare function mergeWithBase(configureTheme: ThemeObject): ThemeObject;

export default baseTheme;
export { mergeWithBase };
