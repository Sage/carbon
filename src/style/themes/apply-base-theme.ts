import baseTheme from "./base";

interface ApplyBaseThemeArgs {
  theme: object;
  [key: string]: unknown;
}

/**
 * To be used by a styled component, this util applies Carbon's base theme to the component if no theme is provided from a parent `ThemeProvider`.
 * Otherwise, it preserves the provided theme.
 *
 * @see {@link https://styled-components.com/docs/api#themeprovider | `ThemeProvider`} docs
 */
export default ({ theme, ...props }: ApplyBaseThemeArgs) => {
  return {
    // A styled component's `theme` prop defaults to an empty object if ThemeProvider is not used.
    // This ensures that the base theme is applied if no theme is provided.
    theme: Object.keys(theme).length === 0 ? baseTheme : theme,
    ...props,
  };
};
