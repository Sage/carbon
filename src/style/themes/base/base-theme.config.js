export default (palette) => {
  const { atOpacity } = palette,

      blackWithOpacity = atOpacity('#000000'),
      baseWithOpacity = atOpacity(palette.productGreen);

  return {
    name: 'base',

    palette,

    spacing: 8,

    space: [0, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80],

    colors: {
      // main
      base: palette.productGreen,
      primary: palette.genericGreenTint(15),
      secondary: palette.genericGreenShade(35),
      tertiary: palette.genericGreenShade(55),
      brand: palette.brilliantGreen,
      disabled: palette.genericGreenTint(40),
      whiteMix: palette.genericGreenTint(90),
      withOpacity: baseWithOpacity(0.55),

      // generic
      black: '#000000',
      slate: '#003349',
      white: '#FFFFFF',

      // element
      border: palette.slateTint(40),
      dashedBorder: '#99ADB6',
      dashedButtonText: 'rgba(0, 0, 0, 0.9)',
      dashedHoverBackground: '#CCD6DB',
      focusedIcon: palette.slateTint(20),
      focusedLinkBackground: palette.goldTint(50),
      previewBackground: palette.slateTint(75),
      hoveredTabKeyline: palette.genericGreenTint(30),

      // status
      error: palette.errorRed,
      focus: palette.gold,
      info: palette.productBlueShade(3),
      success: palette.brilliantGreenShade(20),
      warning: palette.carrotOrange,
      destructive: {
        hover: palette.errorRedShade(20)
      },
      asterisk: palette.errorRed
    },

    anchorNavigation: {
      divider: palette.slateTint(80),
      navItemHoverBackground: palette.slateTint(90)
    },

    accordion: {
      border: palette.slateTint(80),
      background: palette.slateTint(90)
    },

    tileSelect: {
      border: palette.slateTint(75),
      disabledBackground: palette.slateTint(90),
      hoverBackground: palette.slateTint(95),
      descriptionColor: blackWithOpacity(0.55),
      disabledText: blackWithOpacity(0.3)
    },

    batchSelection: {
      lightTheme: palette.slateTint(70)
    },

    hr: {
      background: palette.slateTint(80)
    },

    editor: {
      border: palette.slateTint(40),
      counter: 'rgba(0,0,0,0.55)',
      placeholder: 'rgba(0,0,0,0.30)',
      button: {
        hover: palette.slateTint(80)
      },
      toolbar: {
        background: palette.slateTint(95)
      }
    },

    menu: {
      focus: palette.slateTint(95),
      divider: palette.slateTint(90),
      itemColor: blackWithOpacity(0.9),
      itemColorDisabled: blackWithOpacity(0.3),

      light: {
        background: palette.slateTint(90),
        selected: palette.slateTint(85),
        divider: palette.slateTint(80)
      },

      dark: {
        divider: palette.slateTint(10),
        submenuBackground: palette.slateShade(50),
        selected: palette.slateTint(10)
      }
    },

    form: {
      invalid: palette.slateTint(95)
    },

    card: {
      footerBackground: palette.slateTint(95),
      footerBorder: palette.slateTint(90)
    },

    carousel: {
      activeSelectorBackground: palette.slateTint(40),
      inactiveSelectorBackground: palette.slateTint(80)
    },

    flatTable: {
      light: {
        headerBackground: palette.slateTint(80),
        border: palette.slateTint(70)
      },

      dark: {
        headerBackground: palette.slateTint(20),
        border: palette.slateTint(40)
      },

      transparentWhite: {
        headerBackground: '#fff',
        border: '#fff'
      },

      transparentBase: {
        headerBackground: palette.slateTint(95),
        border: palette.slateTint(95)
      },

      drawerSidebar: {
        headerBackground: palette.slateTint(93),
        hover: palette.slateTint(85),
        highlighted: palette.slateTint(80),
        selected: palette.slateTint(70)
      },

      hover: palette.slateTint(95),
      headerIconColor: palette.slateTint(60),
      selected: palette.slateTint(85),
      highlighted: palette.slateTint(90)
    },

    help: {
      color: blackWithOpacity(0.65),
      hover: blackWithOpacity(0.9)
    },

    pod: {
      border: palette.slateTint(80),
      secondaryBackground: palette.slateTint(95),
      tertiaryBackground: palette.slateTint(90),
      tileBackground: '#FFFFFF',
      footerBackground: palette.slateTint(95),
      hoverBackground: palette.slateTint(85)
    },

    text: {
      color: blackWithOpacity(0.9),
      placeholder: blackWithOpacity(0.3),
      size: '14px'
    },

    readOnly: {
      textboxBackground: palette.slateTint(98),
      textboxBorder: palette.slateTint(80),
      textboxText: blackWithOpacity(0.74)
    },

    definitionList: {
      dtTextDark: blackWithOpacity(0.9),
      dtTextLight: blackWithOpacity(0.65),
      ddText: blackWithOpacity(0.65)
    },

    disabled: {
      border: palette.slateTint(80),
      button: palette.slateTint(90),
      disabled: blackWithOpacity(0.55),
      input: palette.slateTint(95),
      text: blackWithOpacity(0.3),
      buttonText: 'rgba(0,0,0,.2)',
      background: palette.slateTint(90),
      switch: palette.slateTint(89.5)
    },

    draggableItem: {
      border: palette.slateTint(90)
    },

    checkable: {
      checked: 'rgba(0,0,0,0.90)'
    },

    table: {
      primary: palette.slateTint(95),
      secondary: palette.slateTint(80),
      tertiary: palette.slateTint(10),
      header: palette.slateTint(20),
      hover: palette.slateTint(90),
      selected: palette.slateTint(85),
      zebra: palette.slateTint(98),
      dragging: palette.slateTint(90)
    },

    drawer: {
      background: palette.slateTint(93),
      divider: palette.slateTint(85)
    },

    pager: {
      active: 'rgba(0,0,0,0.90)',
      disabled: 'rgba(0,0,0,0.3)'
    },

    icon: {
      default: 'rgba(0,0,0,0.65)',
      defaultHover: 'rgba(0,0,0,0.90)',
      onLightBackground: palette.slateTint(40),
      onLightBackgroundHover: palette.slateTint(20),
      disabled: 'rgba(0,0,0,0.30)'
    },

    popoverContainer: {
      iconColor: 'rgba(0,0,0,0.90)'
    },

    navigationBar: {
      light: {
        background: palette.slateTint(90),
        borderBottom: palette.slateTint(85)
      },

      dark: {
        background: '#003349',
        borderBottom: '#003349'
      }
    },

    numeralDate: {
      passive: '#668592',
      error: '#C7384F'
    },

    portrait: {
      border: palette.slateTint(50),
      background: palette.slateTint(95),
      initials: 'rgba(0,0,0,0.65)'
    },

    pill: {
      neutral: palette.slateTint(30),
      warning: palette.carrotOrangeTint(20),
      neutralBackgroundFocus: palette.slateTint(10),
      warningButtonFocus: palette.carrotOrange,
      errorButtonFocus: palette.errorRedShade(20)
    },

    search: {
      active: palette.goldTint(50),
      button: '#255BC7',
      passive: palette.slateTint(80),
      searchActive: '#668592'
    },

    select: {
      border: '#bfccd2',
      selected: palette.slateTint(95)
    },

    shadows: {
      depth1: '0 5px 5px 0 rgba(0,20,29,0.2), 0 10px 10px 0 rgba(0,20,29,0.1)',
      depth2: '0 10px 20px 0 rgba(0,20,29,0.2), 0 20px 40px 0 rgba(0,20,29,0.1)',
      depth3: '0 10px 30px 0 rgba(0,20,29,0.1), 0 30px 60px 0 rgba(0,20,29,0.1)',
      depth4: '0 10px 40px 0 rgba(0,20,29,0.04), 0 50px 80px 0 rgba(0,20,29,0.1)',
      /* no space after comma as it is stripped from the variable when used in the spec */
      cards: '0 3px 3px 0 rgba(0,20,29,0.2),0 2px 4px 0 rgba(0,20,29,0.15)',
      cardsIE: '0 3px 3px 0 rgba(0,20,29,0.2),0 2px 4px 0 rgba(0,20,29,0.15), 0 0 1px 0 rgba(0,20,29,0.15)'
    },

    switch: {
      off: '#CCD6DB'
    },

    tile: {
      border: palette.slateTint(80),
      separator: palette.slateTint(90)
    },

    tab: {
      background: palette.slateTint(80),
      altHover: palette.slateTint(85)
    },

    note: {
      timeStamp: 'rgba(0,0,0,0.65)'
    },

    zIndex: {
      overlay: 1000,
      popover: 2000,
      modal: 3000,
      header: 4000,
      fullScreenModal: 5000,
      notification: 6000
    }
  };
};
