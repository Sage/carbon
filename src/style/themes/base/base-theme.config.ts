import tokens from "@sage/design-tokens/js/base/common";

import { PaletteFunction } from "../../palette";

export type BasePalette = Record<string, string | PaletteFunction> & {
  blackOpacity: PaletteFunction;
  whiteOpacity: PaletteFunction;
  [key: string]: PaletteFunction;
};

export interface Colors {
  primary: string;
  secondary: string;
  tertiary: string;
  loadingBarBackground: string;

  // generic
  white: "#FFFFFF";

  // status
  error: string;
  focus: string;
  info: string;
  warning: string;
  destructive: {
    hover: string;
  };
  warningText: string;
  placeholder: string;
}

export interface ThemeObject extends Record<string, unknown> {
  name: string;
  palette: BasePalette;
  spacing: number;
  space: string[];

  colors: Partial<Colors>;

  disabled: {
    background: string;
  };

  zIndex: {
    smallOverlay: number;
    overlay: number;
    popover: number;
    nav: number;
    globalNav: number;
    modal: number;
    header: number;
    fullScreenModal: number;
    notification: number;
    aboveAll: number;
  };

  compatibility: Record<string, string | undefined>;
}

export const space = [
  "var(--spacing000)",
  "var(--spacing100)",
  "var(--spacing200)",
  "var(--spacing300)",
  "var(--spacing400)",
  "var(--spacing500)",
  "var(--spacing600)",
  "var(--spacing700)",
  "var(--spacing800)",
  "var(--spacing900)",
  "var(--spacing1000)",
];

export default (palette: BasePalette): ThemeObject => {
  return {
    name: "base",

    palette,

    spacing: 8,

    space,

    colors: {
      // main
      primary: palette.genericGreenShade(15),
      secondary: palette.genericGreenShade(35),
      tertiary: palette.genericGreenShade(55),
      loadingBarBackground: palette.genericGreenTint(70),

      // generic
      white: "#FFFFFF",

      // status
      error: palette.errorRed as string,
      focus: palette.gold as string,
      info: palette.productBlueShade(3),
      warning: palette.carrotOrange as string,
      warningText: palette.carrotOrangeShade(20),
      destructive: {
        hover: palette.errorRedShade(20),
      },
      placeholder: palette.blackOpacity(0.55),
    },

    disabled: {
      background: palette.slateTint(90),
    },

    zIndex: {
      smallOverlay: 10,
      overlay: 1000,
      nav: 2998,
      globalNav: 2999,
      modal: 3000,
      header: 4000,
      fullScreenModal: 5000,
      popover: 6000,
      notification: 7000,
      aboveAll: 9999,
    },

    get compatibility() {
      return {
        ...tokens,

        colorsActionMajor150: this.colors.loadingBarBackground,
        colorsActionMajor500: this.colors.primary,
        colorsActionMajor600: this.colors.secondary,

        colorsActionDisabled500: this.disabled.background,

        colorsSemanticFocus500: this.colors.focus,

        colorsSemanticNegative500: this.colors.error,
        colorsSemanticNegative600: this?.colors?.destructive?.hover,

        colorsSemanticCaution500: this.colors.warning,

        colorsSemanticInfo500: this.colors.info,
      };
    },
  };
};
