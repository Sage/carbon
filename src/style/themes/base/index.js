import configureBase from './base-theme.config';
import baseColors from '../../color-config';
import atOpacity from '../../utils/at-opacity';
import { mergeDeep } from '../../utils/merge-deep';
import generatePalette from '../../palette';
import addHexSymbols from '../../utils/add-hex-symbols';

const palette = {
  ...generatePalette(baseColors),
  atOpacity,
  ...addHexSymbols(baseColors)
};


const baseTheme = configureBase(palette);

export default baseTheme;

export const mergeWithBase = (configureTheme) => {
  const themeToMergeWithBase = configureTheme(palette);

  return {
    ...mergeDeep(baseTheme, themeToMergeWithBase),
    palette
  };
};
