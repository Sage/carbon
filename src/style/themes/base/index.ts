import configureBase from "./base-theme.config";
import baseColors from "../../color-config";
import atOpacity from "../../utils/at-opacity";
import { mergeDeep } from "../../utils/merge-deep";
import generatePalette from "../../palette";
import addHexSymbols from "../../utils/add-hex-symbols";
import type { BasePalette, ThemeObject } from "../theme.types";

const colors = generatePalette(baseColors);
const colorsWithHex = addHexSymbols(baseColors);
const blackOpacity = atOpacity("#000000");
const whiteOpacity = atOpacity("#FFFFFF");

const palette: BasePalette = {
  ...colors,
  ...colorsWithHex,
  blackOpacity,
  whiteOpacity,
};

const baseTheme = configureBase(palette);

export default baseTheme;

export const mergeWithBase = (
  configureTheme: (p: BasePalette) => Partial<ThemeObject>,
): Partial<ThemeObject> => {
  const themeToMergeWithBase = configureTheme(palette);

  return {
    ...mergeDeep(baseTheme, themeToMergeWithBase),
    palette,
  };
};
