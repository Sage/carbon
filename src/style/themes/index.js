// eslint-disable-next-line import/prefer-default-export
export const THEMES = {
  base: 'base',
  none: 'none',
  classic: 'classic',
  small: 'small',
  medium: 'medium',
  large: 'large'
};
