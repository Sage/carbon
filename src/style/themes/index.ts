import sageTheme from "./sage";
import mintTheme from "./mint";
import aegeanTheme from "./aegean";

export { sageTheme, mintTheme, aegeanTheme };
export { default as baseTheme } from "./base";
export { default as noTheme } from "./none";

export type { ThemeObject } from "./base";
export const carbonThemeList = [mintTheme, aegeanTheme, sageTheme];
