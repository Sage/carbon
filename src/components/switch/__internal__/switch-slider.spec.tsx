import React from "react";
import TestRenderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import { mount } from "enzyme";

import {
  assertStyleMatch,
  carbonThemesJestTable,
} from "../../../__spec_helper__/test-utils";
import Loader from "../../loader/loader.component";
import SwitchSlider, { SwitchSliderProps } from "./switch-slider.component";
import SwitchSliderPanel from "./switch-slider-panel.style";
import { ThemeObject } from "../../../style/themes/base";
import CarbonProvider from "../../carbon-provider/carbon-provider.component";
import StyledSwitchSlider from "./switch-slider.style";

function render(props?: Partial<SwitchSliderProps>) {
  return TestRenderer.create(<SwitchSlider {...props} />);
}

function renderWithTheme(
  props: Partial<SwitchSliderProps>,
  theme?: string | Partial<ThemeObject>,
  renderer = TestRenderer.create
) {
  return renderer(
    <ThemeProvider theme={theme}>
      <SwitchSlider {...props} />
    </ThemeProvider>
  );
}

describe("SwitchSlider", () => {
  describe("base theme", () => {
    it("renders as expected", () => {
      expect(render()).toMatchSnapshot();
    });

    describe("Panel content", () => {
      describe("default", () => {
        const panels = render().root.findAllByType(SwitchSliderPanel);

        it('renders the text "OFF" in the panel', () => {
          expect(panels[0].props.children).toBe("OFF");
        });

        it("renders only one panel", () => {
          expect(panels.length).toBe(1);
        });
      });

      describe("when checked=true", () => {
        const panels = render({ checked: true }).root.findAllByType(
          SwitchSliderPanel
        );

        it('renders the text "ON" in the panel', () => {
          expect(panels[0].props.children).toBe("ON");
        });

        it("renders only one panel", () => {
          expect(panels.length).toBe(1);
        });
      });

      describe("when loading=true", () => {
        const panels = render({ loading: true }).root.findAllByType(
          SwitchSliderPanel
        );

        it("renders a Loader in the first panel", () => {
          expect(panels[0].props.children.type).toBe(Loader);
        });

        it("renders only one panel", () => {
          expect(panels.length).toBe(1);
        });
      });
    });

    describe("when checked=true", () => {
      const wrapper = render({ checked: true }).toJSON();

      it("applies the correct base styles", () => {
        assertStyleMatch(
          {
            backgroundColor: "var(--colorsActionMinor500)",
          },
          wrapper
        );
      });

      it("applies the correct ::before styles", () => {
        assertStyleMatch(
          {
            transform: "translateX(36px)",
          },
          wrapper,
          { modifier: "::before" }
        );
      });
    });

    describe("when disabled=true", () => {
      const wrapper = render({ disabled: true }).toJSON();

      it("applies the correct base styles", () => {
        assertStyleMatch(
          {
            borderColor: "var(--colorsActionDisabled600)",
          },
          wrapper
        );
      });

      it("applies the correct ::before styles", () => {
        assertStyleMatch(
          {
            backgroundColor: "var(--colorsActionDisabled600)",
          },
          wrapper,
          { modifier: "::before" }
        );
      });

      it("applies the correct SwitchSliderPanel styles", () => {
        assertStyleMatch(
          {
            color: "var(--colorsUtilityYin030)",
          },
          wrapper,
          {
            modifier: `${SwitchSliderPanel}`,
          }
        );
      });
    });

    describe("when checked=true && disabled=true", () => {
      const wrapper = render({ checked: true, disabled: true }).toJSON();

      it("applies the correct SwitchSliderPanel styles", () => {
        assertStyleMatch(
          {
            color: "var(--colorsUtilityYin030)",
          },
          wrapper,
          {
            modifier: `${SwitchSliderPanel}`,
          }
        );
      });
    });

    describe("when size=large", () => {
      describe("default", () => {
        const wrapper = render({ size: "large" }).toJSON();

        it("applies the correct ::before styles", () => {
          assertStyleMatch(
            {
              height: "32px",
              width: "32px",
            },
            wrapper,
            { modifier: "::before" }
          );
        });
      });

      describe("and checked=true", () => {
        const wrapper = render({ checked: true, size: "large" }).toJSON();

        it("applies the correct ::before styles", () => {
          assertStyleMatch(
            {
              transform: "translateX(38px)",
            },
            wrapper,
            { modifier: "::before" }
          );
        });
      });
    });
  });

  describe.each(carbonThemesJestTable)(
    "when the theme is set to %s",
    (themeName, theme) => {
      describe("default", () => {
        const wrapper = renderWithTheme({}, theme).toJSON();

        it("applies the correct base styles", () => {
          assertStyleMatch(
            {
              borderColor: "var(--colorsActionMinor400)",
            },
            wrapper
          );
        });

        it("applies the correct ::before styles", () => {
          assertStyleMatch(
            {
              backgroundColor: "var(--colorsActionMinor400)",
            },
            wrapper,
            { modifier: "::before" }
          );
        });
      });

      describe("and checked=true", () => {
        const wrapper = renderWithTheme({ checked: true }, theme).toJSON();

        it("applies the correct base styles", () => {
          assertStyleMatch(
            {
              backgroundColor: "var(--colorsActionMinor500)",
            },
            wrapper
          );
        });
      });

      describe("and disabled=true", () => {
        const wrapper = renderWithTheme({ disabled: true }, theme).toJSON();

        it("applies the correct base styles", () => {
          assertStyleMatch(
            {
              borderColor: "var(--colorsActionDisabled600)",
            },
            wrapper
          );
        });

        it("applies the correct SwitchSliderPanel styles", () => {
          assertStyleMatch(
            {
              color: "var(--colorsUtilityYin030)",
            },
            wrapper,
            {
              modifier: `${SwitchSliderPanel}`,
            }
          );
        });
      });

      describe("when checked=true && disabled=true", () => {
        const wrapper = renderWithTheme(
          { checked: true, disabled: true },
          theme
        ).toJSON();

        it("applies the correct base styles", () => {
          assertStyleMatch(
            {
              backgroundColor: "var(--colorsActionDisabled500)",
            },
            wrapper
          );
        });

        it("applies the correct SwitchSliderPanel styles", () => {
          assertStyleMatch(
            {
              color: "var(--colorsUtilityYin030)",
            },
            wrapper,
            {
              modifier: `${SwitchSliderPanel}`,
            }
          );
        });
      });
    }
  );

  describe("rounded corners", () => {
    it.each<SwitchSliderProps["size"]>(["small", "large"])(
      "has the expected border radius styling when size is %s",
      (size) => {
        const wrapper = render({ size }).toJSON();
        assertStyleMatch(
          {
            borderRadius: "var(--borderRadius400)",
          },
          wrapper
        );
      }
    );

    it.each<SwitchSliderProps["size"]>(["small", "large"])(
      "has the expected border radius styling when size is %s and roundedCornersOptOut is true",
      (size) => {
        const wrapper = mount(
          <CarbonProvider roundedCornersOptOut>
            <SwitchSlider size={size} />
          </CarbonProvider>
        );
        assertStyleMatch(
          {
            borderRadius: size === "large" ? "30px" : "90px",
          },
          wrapper.find(StyledSwitchSlider)
        );
      }
    );
  });
});
