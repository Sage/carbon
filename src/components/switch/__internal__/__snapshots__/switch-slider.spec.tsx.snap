// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchSlider base theme renders as expected 1`] = `
.c2 {
  border: 0;
  color: var(--colorsActionMinorYang100);
  margin: auto;
  margin-top: 5px;
}

.c2[type="on"] {
  margin-left: 9px;
}

.c2[type="off"] {
  color: var(--colorsActionMinor500);
  margin-right: 6px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  font-weight: bold;
  height: 24px;
  left: 0;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  position: absolute;
  text-transform: uppercase;
  top: 0;
  width: 60px;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  z-index: 2;
  border-radius: 90px;
  border-style: solid;
  border-color: var(--colorsActionMinor400);
  border-width: var(--borderWidth200);
}

.c0::before {
  background-color: var(--colorsActionMinor400);
  bottom: 4px;
  content: "";
  height: 16px;
  position: absolute;
  left: 4px;
  -webkit-transition: -webkit-transform 0.4s;
  -webkit-transition: transform 0.4s;
  transition: transform 0.4s;
  width: 16px;
  z-index: 1;
  border-radius: 50%;
}

.c0 .c5 {
  position: absolute;
  right: -30px;
  height: 100%;
}

.c3 .c1 {
  color: var(--colorsUtilityYin030);
}

.c4 .c1 {
  color: var(--colorsUtilityYin030);
}

.c4 .c1 {
  color: var(--colorsUtilityYin030);
}

<span
  className="c0"
>
  <div
    className="c1 c2"
    type="off"
  >
    OFF
  </div>
</span>
`;
